### PR TITLE
move decode_raw

### PR DESCRIPTION
**Pull request by whit537**
It doesn't belong in resources, it belongs down in simplates.

### CODE REVIEW

**Comment by whit537** [comment]
Includes #490.

**Comment by whit537** [comment]
Rebased on master.

**Comment by whit537** [comment]
Okay, I've got a clean 2.6 env locally, and am seeing 15 failures.

**Comment by whit537** [comment]
So the difference introduced is that Dynamic now takes a bytestring instead of unicode. What changed between 2.6 and 2.7 around that?

**Comment by whit537** [comment]
Ah, sure: in the test suite we're passing in a unicode instead of a bytestring. But why doesn't that fail under 2.7?

**Comment by whit537** [comment]
Yeeeeeeeeeeeeeeeikes.

```
(Pdb) self.raw
u'[---]\n[---] text/plain via stdlib_template\n'
(Pdb) self.decoded
u'[\x00-\x00-\x00-\x00]\x00\n\x00[\x00-\x00-\x00-\x00]\x00 \x00t\x00e\x00x\x00t\x00/\x00p\x00l\x00a\x00i\x00n\x00 \x00v\x00i\x00a\x00 \x00s\x00t\x00d\x00l\x00i\x00b\x00_\x00t\x00e\x00m\x00p\x00l\x00a\x00t\x00e\x00\n\x00'
(Pdb)
```

Are those null bytes?

**Comment by whit537** [comment]
Under 2.7:

```
(Pdb) self.raw
u'[---]\n[---] text/plain via stdlib_template\n'
(Pdb) self.decoded
u'[---]\n[---] text/plain via stdlib_template\n'
(Pdb)
```

**Comment by whit537** [comment]
A difference in StringIO?

**Comment by pjz** [comment]
There's also the nuclear option: it is permissible, in 2015, to drop support for python 2.6, since python 2.7 has been out for _5 years_.  Even debian no longer ships with 2.6.

**Comment by whit537** [comment]
@pjz Fair enough. This isn't too hard to work around though.

I found a reference to a similar problem: http://stackoverflow.com/questions/8339470/stringio-and-pystache-generate-spurious-null-bytes. I haven't found a Python bug that was fixed in 2.7 though.

Workaround forthcoming ...

**Comment by Changaco** [comment]
+1 to dropping support for 2.6 and closing #486

**Comment by whit537** [comment]
Dropping support for 2.6 reticketed as #492.

Ready for review.

File: aspen/simplates/__init__.py [hunk 1]
```diff
@@ -30,6 +31,53 @@ def _ordinal(n):
     return str(n)
 
 
+def _decode(raw):
+    """As per PEP 263, decode raw data according to the encoding specified in
+       the first couple lines of the data, or in ASCII.  Non-ASCII data without
+       an encoding specified will cause UnicodeDecodeError to be raised.
+    """
+    assert type(raw) is (bytes if sys.version_info[0] == 3 else str)  # sanity check
```
Comment (whit537) [inline aspen/simplates/__init__.py]: I mean, not that we've ported Aspen to Python 3 yet ... :P

**Pull request merged**